File: ts/packages/server/src/datasource/policy.ts
import { DataSourceErrorCode, DataSourceExecutionError } from "./types.js";

export interface DataSourceTableConfig {
  name: string;
  tenantColumn?: string;
}

const blockedSqlKeywords = [
  "alter",
  "analyze",
  "begin",
  "call",
  "commit",
  "copy",
  "create",
  "delete",
  "drop",
  "export",
  "grant",
  "import",
  "insert",
  "load",
  "merge",
  "revoke",
  "rollback",
  "set",
  "truncate",
  "update",
  "vacuum",
];

export function isSingleReadOnlyStatement(query: string): boolean {
  const normalized = trimSql(query);
  if (!normalized || normalized.includes(";")) {
    return false;
  }
  const lower = normalized.toLowerCase();
  if (!/^(select|with)\s/i.test(normalized)) {
    return false;
  }
  return !blockedSqlKeywords.some((keyword) => containsIdentifier(lower, keyword));
}

export function containsIdentifier(query: string, identifier: string): boolean {
  const escaped = identifier.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`(^|[^A-Za-z0-9_])${escaped}([^A-Za-z0-9_]|$)`, "i").test(query);
}

export function referencedTables(
  query: string,
  explicitTableNames: readonly string[] = [],
  tables: readonly DataSourceTableConfig[] = []
): string[] {
  if (explicitTableNames.length > 0) {
    return [...explicitTableNames];
  }
  return tables
    .filter((table) => table.name && containsIdentifier(query, table.name))
    .map((table) => table.name);
}

export function validateReadOnlyQuery(
  query: string,
  explicitTableNames: readonly string[] = [],
  tables: readonly DataSourceTableConfig[] = []
): void {
  if (!query.trim()) {
    throw new DataSourceExecutionError({
      code: DataSourceErrorCode.InvalidArgument,
      message: "query is required",
    });
  }
  if (!isSingleReadOnlyStatement(query)) {
    throw new DataSourceExecutionError({
      code: DataSourceErrorCode.QueryInvalid,
      message: "query must be a single read-only SELECT statement",
    });
  }
  if (tables.length === 0) {
    return;
  }

  const allowedTables = new Set(tables.map((table) => table.name.toLowerCase()));
  const tableNames = referencedTables(query, explicitTableNames, tables);
  if (tableNames.length === 0) {
    throw new DataSourceExecutionError({
      code: DataSourceErrorCode.QueryInvalid,
      message: "query must reference a supported datasource table",
    });
  }
  for (const tableName of tableNames) {
    if (!allowedTables.has(tableName.toLowerCase())) {
      throw new DataSourceExecutionError({
        code: DataSourceErrorCode.TableNotFound,
        message: `unsupported datasource table: ${tableName}`,
      });
    }
  }
}

export function queryWithRowLimit(query: string, rowLimit: number | undefined): string {
  const normalized = trimSql(query);
  const limit = Math.trunc(rowLimit ?? 0);
  if (limit <= 0) {
    return normalized;
  }
  return `SELECT * FROM (${normalized}) AS datasource_query LIMIT ${limit}`;
}

export class DataSourceByteLimitTracker {
  private sentBytes = 0;

  constructor(private readonly byteLimit: number | undefined) {}

  reserve(size: number): void {
    const limit = Math.trunc(this.byteLimit ?? 0);
    if (limit <= 0) {
      return;
    }
    this.sentBytes += size;
    if (this.sentBytes > limit) {
      throw new DataSourceExecutionError({
        code: DataSourceErrorCode.LimitExceeded,
        message: "datasource query byte limit exceeded",
      });
    }
  }
}

function trimSql(query: string): string {
  return query.trim().replace(/;+$/, "").trim();
}

File: go/datasource/policy.go
package datasource

import (
	"fmt"
	"regexp"
	"strings"
)

var readOnlyStatementStartPattern = regexp.MustCompile(`(?i)^(select|with)\s`)

type TableConfig struct {
	Name         string
	TenantColumn string
}

func IsSingleReadOnlyStatement(query string) bool {
	normalized := strings.TrimSpace(query)
	normalized = strings.TrimSuffix(normalized, ";")
	if normalized == "" || strings.Contains(normalized, ";") {
		return false
	}
	if !readOnlyStatementStartPattern.MatchString(normalized) {
		return false
	}
	lower := strings.ToLower(normalized)
	for _, keyword := range []string{
		"alter", "analyze", "begin", "call", "commit", "copy", "create", "delete",
		"drop", "export", "grant", "import", "insert", "load", "merge", "revoke",
		"rollback", "set", "truncate", "update", "vacuum",
	} {
		if ContainsIdentifier(lower, keyword) {
			return false
		}
	}
	return true
}

func ContainsIdentifier(query string, identifier string) bool {
	pattern := regexp.MustCompile(`(?i)(^|[^A-Za-z0-9_])` + regexp.QuoteMeta(identifier) + `([^A-Za-z0-9_]|$)`)
	return pattern.MatchString(query)
}

func ReferencedTables(query string, explicitTableNames []string, tables []TableConfig) []string {
	if len(explicitTableNames) > 0 {
		return append([]string(nil), explicitTableNames...)
	}
	result := []string{}
	for _, table := range tables {
		if table.Name != "" && ContainsIdentifier(query, table.Name) {
			result = append(result, table.Name)
		}
	}
	return result
}

func ValidateReadOnlyQuery(query string, explicitTableNames []string, tables []TableConfig) error {
	if strings.TrimSpace(query) == "" {
		return fmt.Errorf("query is required")
	}
	if !IsSingleReadOnlyStatement(query) {
		return fmt.Errorf("query must be a single read-only SELECT statement")
	}
	if len(tables) == 0 {
		return nil
	}

	tableMap := make(map[string]struct{}, len(tables))
	for _, table := range tables {
		tableName := strings.TrimSpace(table.Name)
		if tableName != "" {
			tableMap[strings.ToLower(tableName)] = struct{}{}
		}
	}
	tableNames := ReferencedTables(query, explicitTableNames, tables)
	if len(tableNames) == 0 {
		return fmt.Errorf("query must reference a supported datasource table")
	}
	for _, tableName := range tableNames {
		if _, ok := tableMap[strings.ToLower(tableName)]; !ok {
			return fmt.Errorf("unsupported datasource table: %s", tableName)
		}
	}
	return nil
}

func QueryWithRowLimit(query string, rowLimit int64) string {
	normalized := strings.TrimSpace(query)
	normalized = strings.TrimSuffix(normalized, ";")
	if rowLimit <= 0 {
		return normalized
	}
	return fmt.Sprintf("SELECT * FROM (%s) AS datasource_query LIMIT %d", normalized, rowLimit)
}
